File: kayak-backend/src/services/method_service.rs
//! 试验方法服务
//!
//! 提供试验方法的业务逻辑

use crate::db::repository::method_repo::MethodRepository;
use crate::db::repository::method_error::MethodRepositoryError;
use crate::models::dto::method_dto::{CreateMethodRequest, MethodDto, MethodListResponse, UpdateMethodRequest};
use serde::Serialize;
use uuid::Uuid;

#[derive(Debug)]
pub enum MethodServiceError {
    Validation(String),
    NotFound,
    Forbidden,
    Repository(MethodRepositoryError),
}

impl From<MethodRepositoryError> for MethodServiceError {
    fn from(err: MethodRepositoryError) -> Self {
        MethodServiceError::Repository(err)
    }
}

/// Validation result for method process definitions (C3 fix: defined in service layer, not handler)
#[derive(Debug, Serialize)]
pub struct ValidationResult {
    pub valid: bool,
    pub errors: Vec<String>,
}

/// Method service trait (M4 fix: defined in service layer, not handler)
#[axum::async_trait]
pub trait MethodServiceTrait: Send + Sync {
    async fn create_method(
        &self,
        request: CreateMethodRequest,
        user_id: Uuid,
    ) -> Result<MethodDto, MethodServiceError>;
    async fn get_method(&self, id: Uuid, user_id: Uuid) -> Result<MethodDto, MethodServiceError>;
    async fn update_method(
        &self,
        id: Uuid,
        request: UpdateMethodRequest,
        user_id: Uuid,
    ) -> Result<MethodDto, MethodServiceError>;
    async fn delete_method(&self, id: Uuid, user_id: Uuid) -> Result<(), MethodServiceError>;
    async fn list_methods(
        &self,
        user_id: Uuid,
        page: i64,
        size: i64,
    ) -> Result<MethodListResponse, MethodServiceError>;
    async fn validate_method(
        &self,
        process_definition: serde_json::Value,
    ) -> Result<ValidationResult, MethodServiceError>;
}

pub struct MethodService<R: MethodRepository> {
    repository: R,
}

impl<R: MethodRepository> MethodService<R> {
    pub fn new(repository: R) -> Self {
        Self { repository }
    }

    /// 创建方法
    pub async fn create_method(
        &self,
        request: CreateMethodRequest,
        user_id: Uuid,
    ) -> Result<MethodDto, MethodServiceError> {
        // 验证请求
        self.validate_create_request(&request)?;

        // 创建实体
        let method = crate::models::entities::Method::new(
            request.name,
            request.description,
            request.process_definition,
            request.parameter_schema,
            user_id,
        );

        // 保存到数据库
        let created = self.repository.create(&method).await?;

        Ok(created.into())
    }

    /// 获取方法 (M1/M2 fix: with ownership check)
    pub async fn get_method(&self, id: Uuid, user_id: Uuid) -> Result<MethodDto, MethodServiceError> {
        let method = self.repository.get_by_id(id).await?;
        match method {
            Some(m) => {
                if m.created_by != user_id {
                    return Err(MethodServiceError::Forbidden);
                }
                Ok(m.into())
            }
            None => Err(MethodServiceError::NotFound),
        }
    }

    /// 更新方法 (M1 fix: with ownership check)
    pub async fn update_method(
        &self,
        id: Uuid,
        request: UpdateMethodRequest,
        user_id: Uuid,
    ) -> Result<MethodDto, MethodServiceError> {
        // 验证请求
        if let Some(ref name) = request.name {
            if name.is_empty() || name.len() > 255 {
                return Err(MethodServiceError::Validation("名称长度必须在1-255之间".to_string()));
            }
        }

        // Ownership check
        let existing = self.repository.get_by_id(id).await?;
        match existing {
            Some(m) if m.created_by != user_id => {
                return Err(MethodServiceError::Forbidden);
            }
            None => return Err(MethodServiceError::NotFound),
            _ => {}
        }

        // m8 fix: Treat empty description string as None for consistency
        let description = request.description.filter(|d| !d.is_empty());

        let updated = self.repository.update(
            id,
            request.name,
            description,
            request.process_definition,
            request.parameter_schema,
        ).await?;

        Ok(updated.into())
    }

    /// 删除方法 (M1 fix: with ownership check)
    pub async fn delete_method(&self, id: Uuid, user_id: Uuid) -> Result<(), MethodServiceError> {
        // Ownership check
        let existing = self.repository.get_by_id(id).await?;
        match existing {
            Some(m) if m.created_by != user_id => {
                return Err(MethodServiceError::Forbidden);
            }
            None => return Err(MethodServiceError::NotFound),
            _ => {}
        }

        self.repository.delete(id).await?;
        Ok(())
    }

    /// 列出用户的方法
    pub async fn list_methods(
        &self,
        user_id: Uuid,
        page: i64,
        size: i64,
    ) -> Result<MethodListResponse, MethodServiceError> {
        let (methods, total) = self.repository.list_by_user(user_id, page, size).await?;

        Ok(MethodListResponse {
            items: methods.into_iter().map(|m| m.into()).collect(),
            total,
            page,
            size,
        })
    }

    /// 验证创建请求
    fn validate_create_request(&self, request: &CreateMethodRequest) -> Result<(), MethodServiceError> {
        // 名称长度验证
        if request.name.is_empty() || request.name.len() > 255 {
            return Err(MethodServiceError::Validation("名称长度必须在1-255之间".to_string()));
        }

        // JSON格式验证
        if !request.process_definition.is_object() {
            return Err(MethodServiceError::Validation("过程定义必须是JSON对象".to_string()));
        }

        if !request.parameter_schema.is_object() {
            return Err(MethodServiceError::Validation("参数Schema必须是JSON对象".to_string()));
        }

        Ok(())
    }

    /// 验证过程定义 (C3 fix: returns service-layer ValidationResult)
    pub fn validate_process_definition(
        &self,
        process_definition: &serde_json::Value,
    ) -> ValidationResult {
        let mut errors = Vec::new();

        // 检查必须是对象
        if !process_definition.is_object() {
            return ValidationResult {
                valid: false,
                errors: vec!["过程定义必须是JSON对象".to_string()],
            };
        }

        // 检查nodes字段
        let nodes = match process_definition.get("nodes") {
            Some(serde_json::Value::Array(nodes)) => nodes,
            Some(_) => {
                errors.push("'nodes'字段必须是数组".to_string());
                return ValidationResult {
                    valid: false,
                    errors,
                };
            }
            None => {
                errors.push("缺少'nodes'字段".to_string());
                return ValidationResult {
                    valid: false,
                    errors,
                };
            }
        };

        // 有效节点类型
        let valid_types = [
            "Start", "Read", "Control", "Delay", "Decision",
            "Branch", "Wait", "Record", "Config", "Subprocess", "End",
        ];

        let mut has_start = false;
        let mut has_end = false;
        let mut node_ids = std::collections::HashSet::new();

        for node in nodes {
            // 检查节点类型
            if let Some(node_type) = node.get("type").and_then(|v| v.as_str()) {
                if !valid_types.contains(&node_type) {
                    let node_id = node.get("id").and_then(|v| v.as_str()).unwrap_or("unknown");
                    errors.push(format!("节点'{}'的类型'{}'无效", node_id, node_type));
                }
                if node_type == "Start" {
                    has_start = true;
                }
                if node_type == "End" {
                    has_end = true;
                }
            } else {
                errors.push("节点缺少'type'字段".to_string());
            }

            // 检查节点ID唯一性
            if let Some(node_id) = node.get("id").and_then(|v| v.as_str()) {
                if !node_ids.insert(node_id.to_string()) {
                    errors.push(format!("节点ID'{}'重复", node_id));
                }
            }
        }

        if !has_start {
            errors.push("缺少Start节点".to_string());
        }
        if !has_end {
            errors.push("缺少End节点".to_string());
        }

        ValidationResult {
            valid: errors.is_empty(),
            errors,
        }
    }
}

File: kayak-frontend/lib/services/method_service.dart
import '../models/common.dart';
import '../models/method.dart';
import 'api_client.dart';

/// MethodService — 试验方法服务
///
/// 封装方法相关的 HTTP API 调用，提供类型安全的接口。
/// 使用 [ApiClient] 发起请求，利用 [ErrorInterceptor] 统一处理错误。
///
/// 所有方法不捕获异常 —— 异常由 ErrorInterceptor 和 Provider 层处理。
class MethodService {
  /// 创建 MethodService 实例
  ///
  /// [client] ApiClient 实例，提供 HTTP GET/POST/PUT/DELETE 方法
  MethodService(this._client);

  final ApiClient _client;

  /// 获取方法列表
  ///
  /// 调用 `GET /api/v1/methods`。
  ///
  /// 返回 [List]<[Method]> 方法列表。
  /// 如果请求失败，抛出 [DioException]。
  Future<List<Method>> list() async {
    final response = await _client.get(
      '/api/v1/methods',
    );

    final apiResponse = ApiResponse<List<Method>>.fromJson(
      response.data as Map<String, dynamic>,
      (json) => (json as List<dynamic>)
          .map((e) => Method.fromJson(e as Map<String, dynamic>))
          .toList(),
    );

    return apiResponse.data;
  }

  /// 根据 ID 获取方法详情
  ///
  /// 调用 `GET /api/v1/methods/{id}`。
  ///
  /// [id] 方法 ID（UUID）
  /// 返回 [Method] 对象
  /// 如果资源不存在，抛出 404 [DioException]
  Future<Method> getById(String id) async {
    final response = await _client.get(
      '/api/v1/methods/$id',
    );

    final apiResponse = ApiResponse<Method>.fromJson(
      response.data as Map<String, dynamic>,
      (json) => Method.fromJson(json as Map<String, dynamic>),
    );

    return apiResponse.data;
  }

  /// 创建方法
  ///
  /// 调用 `POST /api/v1/methods`。
  ///
  /// [data] 包含 name, description, processDefinition, parameterSchema 的请求体
  /// 返回创建的 [Method] 对象
  Future<Method> create(Map<String, dynamic> data) async {
    final response = await _client.post(
      '/api/v1/methods',
      data: data,
    );

    final apiResponse = ApiResponse<Method>.fromJson(
      response.data as Map<String, dynamic>,
      (json) => Method.fromJson(json as Map<String, dynamic>),
    );

    return apiResponse.data;
  }

  /// 更新方法
  ///
  /// 调用 `PUT /api/v1/methods/{id}`。
  /// 支持部分更新：只传递需要变更的字段。
  ///
  /// [id] 方法 ID
  /// [data] 需要更新的字段
  /// 返回更新后的 [Method] 对象
  Future<Method> update(String id, Map<String, dynamic> data) async {
    final response = await _client.put(
      '/api/v1/methods/$id',
      data: data,
    );

    final apiResponse = ApiResponse<Method>.fromJson(
      response.data as Map<String, dynamic>,
      (json) => Method.fromJson(json as Map<String, dynamic>),
    );

    return apiResponse.data;
  }

  /// 删除方法
  ///
  /// 调用 `DELETE /api/v1/methods/{id}`。
  ///
  /// [id] 方法 ID
  Future<void> delete(String id) async {
    await _client.delete(
      '/api/v1/methods/$id',
    );
  }

  /// 验证方法定义
  ///
  /// 调用 `POST /api/v1/methods/validate`。
  ///
  /// [processDefinition] 方法过程定义 JSON
  /// 返回 [ValidationResult] 包含验证结果
  Future<ValidationResult> validate(
    Map<String, dynamic> processDefinition,
  ) async {
    final response = await _client.post(
      '/api/v1/methods/validate',
      data: {'process_definition': processDefinition},
    );

    final apiResponse = ApiResponse<ValidationResult>.fromJson(
      response.data as Map<String, dynamic>,
      (json) =>
          ValidationResult.fromJson(json as Map<String, dynamic>),
    );

    return apiResponse.data;
  }
}

/// 验证结果模型
///
/// 简单的 2 字段响应对象。注意：如果未来增加更多字段，
/// 建议迁移到 freezed + json_serializable 以保持与项目中其他模型一致。
class ValidationResult {
  const ValidationResult({
    required this.valid,
    this.errors = const [],
  });

  factory ValidationResult.fromJson(Map<String, dynamic> json) {
    return ValidationResult(
      valid: json['valid'] as bool? ?? false,
      errors: (json['errors'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
    );
  }

  final bool valid;
  final List<String> errors;

  Map<String, dynamic> toJson() => {
        'valid': valid,
        'errors': errors,
      };
}
